src/contracts/services/silkBasket.ts

```diff
@@ -0,0 +1,328 @@
+
+  // check if data exists
+  if (!silkBasketResponse) {
+    throw new Error('Silk basket response not found in query router response');
+  }
+  // handle error state
+  if (silkBasketResponse.status === BatchItemResponseStatus.ERROR) {
+    throw new Error(silkBasketResponse.response);
+  }
+
+  const pricesResponse = batchQueryResponse.find((
+    responseItem,
+  ) => responseItem.id === SilkBasketBatchResponseItem.PRICES);
+
+  // check if data exists
+  if (!pricesResponse) {
+    throw new Error('prices response not found in query router response');
+  }
+
+  // for price error state we will not throw an error like we did with the basket
+  // data as there is still useful data to get out of the peg info that we would like to parse.
+  // Instead, we will pass the full batch data into the parser (which includes the
+  // error status) and handle the error downstream.
+
+  return parseSilkBasketAndPricesResponse({
+    silkBasketResponse: silkBasketResponse.response,
+    batchBasketPricesResponse: pricesResponse,
+    silkBasketResponseBlockHeight: silkBasketResponse.blockHeight,
+    basketPricesResponseBlockHeight: pricesResponse.blockHeight,
+  });
+}
+
+/**
+ * query the silk basket data
+ */
+function querySilkBasket$({
+  queryRouterContractAddress,
+  queryRouterCodeHash,
+  lcdEndpoint,
+  chainId,
+  oracleContractAddress,
+  oracleCodeHash,
+  silkBasketExpectedOracleKeys = [], // if keys are known in advance it will
+  // increase the efficiency of the response, but even if keys are unknown or
+  // incorrectly passed in compared to the state of the basket,
+  // this service will self-correct using a retry on the actual keys found.
+  silkIndexOracleContractAddress,
+  silkIndexOracleCodeHash,
+  minBlockHeightValidationOptions,
+}:{
+  queryRouterContractAddress: string,
+  queryRouterCodeHash?: string,
+  lcdEndpoint?: string,
+  chainId?: string,
+  oracleContractAddress: string,
+  oracleCodeHash: string,
+  silkBasketExpectedOracleKeys?: string[]
+  silkIndexOracleContractAddress: string,
+  silkIndexOracleCodeHash: string,
+  minBlockHeightValidationOptions?: MinBlockHeightValidationOptions,
+}) {
+  const basketQueryMsg :BatchQueryParams = {
+    id: SilkBasketBatchResponseItem.BASKET,
+    contract: {
+      address: silkIndexOracleContractAddress,
+      codeHash: silkIndexOracleCodeHash,
+    },
+    queryMsg: msgGetSilkBasket(),
+  };
+
+  const pricesQueryMsg:BatchQueryParams = {
+    id: SilkBasketBatchResponseItem.PRICES,
+    contract: {
+      address: oracleContractAddress,
+      codeHash: oracleCodeHash,
+    },
+    queryMsg: msgQueryOraclePrices(silkBasketExpectedOracleKeys),
+  };
+
+  return batchQuery$({
+    contractAddress: queryRouterContractAddress,
+    codeHash: queryRouterCodeHash,
+    lcdEndpoint,
+    chainId,
+    queries: [basketQueryMsg, pricesQueryMsg],
+    minBlockHeightValidationOptions,
+  }).pipe(
+    map(parseSilkBasketAndPricesResponseFromQueryRouter),
+    switchMap((data) => {
```

Comment:
Wouldn't this be better with a try-catch type of setup?

Reply:
better is probably subjective. I don't really see it as an error, because nothing really went wrong, service is operating as expected under those conditions using the output of query1 to feed query2.  You're maybe alluding to the design choice of adding that parsing status type, and that is maybe a bit confusing.

Reply:
And I guess to add one more thing, if we did want to do it as an error, then I would need to pass parameters for query #2 back through the error message and I don't necessarily love that solution either.